File: app/src/main/java/com/example/boardgame/OptionsActivity.java
package com.example.boardgame;

import android.content.Intent;
import android.os.Bundle;
import android.widget.Button;
import android.widget.EditText;
import android.widget.TextView;
import android.widget.Toast;

import androidx.appcompat.app.AppCompatActivity;

public class OptionsActivity extends AppCompatActivity {
    private TextView nicknameValueText;
    private EditText serverUrlInput;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_options);

        nicknameValueText = findViewById(R.id.nicknameValueText);
        serverUrlInput = findViewById(R.id.serverUrlInput);

        Button changeNicknameButton = findViewById(R.id.changeNicknameButton);
        Button saveServerUrlButton = findViewById(R.id.saveServerUrlButton);
        Button useLanServerButton = findViewById(R.id.useLanServerButton);
        Button useWanServerButton = findViewById(R.id.useWanServerButton);
        Button closeButton = findViewById(R.id.closeOptionsButton);

        changeNicknameButton.setOnClickListener(
                view -> startActivity(new Intent(this, NicknameActivity.class))
        );
        closeButton.setOnClickListener(view -> finish());

        saveServerUrlButton.setOnClickListener(view -> saveServerUrlFromInput());
        useLanServerButton.setOnClickListener(view -> {
            ServerSession.useDefaultLan(this);
            renderServerUrl();
            Toast.makeText(this, R.string.options_server_url_saved, Toast.LENGTH_SHORT).show();
        });
        useWanServerButton.setOnClickListener(view -> {
            ServerSession.useDefaultWan(this);
            renderServerUrl();
            Toast.makeText(this, R.string.options_server_url_saved, Toast.LENGTH_SHORT).show();
        });
    }

    @Override
    protected void onResume() {
        super.onResume();
        renderNickname();
        renderServerUrl();
    }

    private void renderNickname() {
        String nickname = SessionPrefs.getNickname(this);
        nicknameValueText.setText(nickname == null || nickname.trim().isEmpty()
                ? getString(R.string.options_nickname_not_set)
                : nickname);
    }

    private void renderServerUrl() {
        serverUrlInput.setText(ServerSession.getServerUrl(this));
    }

    private void saveServerUrlFromInput() {
        String wsUrl = serverUrlInput.getText().toString().trim();
        if (!(wsUrl.startsWith("ws://") || wsUrl.startsWith("wss://"))) {
            serverUrlInput.setError(getString(R.string.options_server_url_invalid));
            return;
        }
        String normalizedUrl = ServerSession.normalizeServerUrl(wsUrl);
        ServerSession.setServerUrl(this, normalizedUrl);
        serverUrlInput.setText(normalizedUrl);
        Toast.makeText(this, R.string.options_server_url_saved, Toast.LENGTH_SHORT).show();
    }
}

File: app/src/main/java/com/example/adchaosdemo/OptionsActivity.kt
package com.example.adchaosdemo

import android.os.Bundle
import android.widget.CompoundButton
import android.widget.Button
import android.widget.CheckBox
import android.widget.EditText
import android.widget.RadioButton
import android.widget.TextView
import android.widget.Toast
import android.widget.Switch
import androidx.appcompat.app.AppCompatActivity

class OptionsActivity : AppCompatActivity() {

    private lateinit var volumeValueText: TextView
    private lateinit var easyVolumeCheck: CheckBox
    private lateinit var soundEffectsSwitch: Switch
    private lateinit var vibrationSwitch: Switch
    private lateinit var diceSpeedNormal: RadioButton
    private lateinit var diceSpeedFast: RadioButton
    private lateinit var debugModeSwitch: Switch
    private lateinit var serverUrlInput: EditText

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_options)

        volumeValueText = findViewById(R.id.volumeValueText)
        easyVolumeCheck = findViewById(R.id.easyVolumeCheck)
        soundEffectsSwitch = findViewById(R.id.soundEffectsSwitch)
        vibrationSwitch = findViewById(R.id.vibrationSwitch)
        diceSpeedNormal = findViewById(R.id.diceSpeedNormal)
        diceSpeedFast = findViewById(R.id.diceSpeedFast)
        debugModeSwitch = findViewById(R.id.debugModeSwitch)
        serverUrlInput = findViewById(R.id.serverUrlInput)

        val minusButton = findViewById<Button>(R.id.volumeMinusButton)
        val plusButton = findViewById<Button>(R.id.volumePlusButton)
        val saveServerUrlButton = findViewById<Button>(R.id.saveServerUrlButton)
        val closeButton = findViewById<Button>(R.id.closeOptionsButton)

        minusButton.setOnClickListener { updateVolumeBy(-1) }
        plusButton.setOnClickListener { updateVolumeBy(1) }
        closeButton.setOnClickListener { finish() }

        easyVolumeCheck.setOnCheckedChangeListener { buttonView, isChecked ->
            if (isChecked) {
                buttonView.isChecked = false
                Toast.makeText(this, R.string.options_easy_volume_locked, Toast.LENGTH_SHORT).show()
            }
        }

        soundEffectsSwitch.setOnCheckedChangeListener { _: CompoundButton, isChecked: Boolean ->
            SessionPrefs.setSoundEffectsOn(this, isChecked)
        }
        vibrationSwitch.setOnCheckedChangeListener { _: CompoundButton, isChecked: Boolean ->
            SessionPrefs.setVibrationOn(this, isChecked)
        }
        diceSpeedNormal.setOnCheckedChangeListener { _: CompoundButton, isChecked: Boolean ->
            if (isChecked) SessionPrefs.setDiceFast(this, false)
        }
        diceSpeedFast.setOnCheckedChangeListener { _: CompoundButton, isChecked: Boolean ->
            if (isChecked) SessionPrefs.setDiceFast(this, true)
        }
        debugModeSwitch.setOnCheckedChangeListener { _: CompoundButton, isChecked: Boolean ->
            SessionPrefs.setDebugMode(this, isChecked)
        }
        saveServerUrlButton.setOnClickListener {
            val wsUrl = serverUrlInput.text.toString().trim()
            if (!(wsUrl.startsWith("ws://") || wsUrl.startsWith("wss://"))) {
                serverUrlInput.error = getString(R.string.options_server_url_invalid)
                return@setOnClickListener
            }
            val normalizedUrl = ServerRoomGateway.normalizeServerUrl(wsUrl)
            ServerRoomGateway.setServerUrl(this, normalizedUrl)
            serverUrlInput.setText(normalizedUrl)
            Toast.makeText(this, R.string.options_server_url_saved, Toast.LENGTH_SHORT).show()
        }
    }

    override fun onResume() {
        super.onResume()
        renderVolume()
        renderExtraOptions()
        renderServerUrl()
        easyVolumeCheck.isChecked = false
    }

    private fun updateVolumeBy(delta: Int) {
        val current = SessionPrefs.getVolume(this)
        SessionPrefs.setVolume(this, current + delta)
        renderVolume()
    }

    private fun renderVolume() {
        val volume = SessionPrefs.getVolume(this)
        volumeValueText.text = getString(R.string.options_volume_value, volume)
    }

    private fun renderExtraOptions() {
        soundEffectsSwitch.isChecked = SessionPrefs.isSoundEffectsOn(this)
        vibrationSwitch.isChecked = SessionPrefs.isVibrationOn(this)
        if (SessionPrefs.isDiceFast(this)) {
            diceSpeedFast.isChecked = true
        } else {
            diceSpeedNormal.isChecked = true
        }
        debugModeSwitch.isChecked = SessionPrefs.isDebugMode(this)
    }

    private fun renderServerUrl() {
        serverUrlInput.setText(ServerRoomGateway.getServerUrl(this))
    }
}
